### PR TITLE
[MC-676][MC-727] Update Domain Extractor in Extractit plugin

### PR DESCRIPTION
**Pull request by dsliwinski-r7**
### Description

Describe the proposed changes:

- support domain extraction from encoded URL
- fix issue where subdomain removal did not work properly in some cases
- improve domain extraction by limiting the number of false positive results

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Run

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "DM1PR09MB1111.namprd09.prod.outlook.com",
        "MN5PR09MB1111.namprd09.prod.outlook.com",
        "BN0PR06CA2222.namprd06.prod.outlook.com",
        "BN1NAM22FT333.eop-nam10.prod.protection.outlook.com",
        "BN0PR06CA2222.outlook.office365.com",
        "bounce.email2.microsoft.com",
        "example.onmicrosoft.com",
        "email2.microsoft.com", 
        "protection.outlook.com",
        "test.email2.microsoft.com",
        "BN0PR06CA2222.mail.protection.outlook.com",
        "click.email2.office.com",
        "leave.email2.office.com",
        "7227040.xt.local",
        "eco.de",
        "atl1s07mta2695.xt.local",
        "BN0PR06CA2222.eop-nam06.prod.protection.outlook.com
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "mx.google.com",
        "drive-notifications.bounces.google.com",
        "egress-00000.ess.de.barracuda.com",
        "mail-aaa-bbbb.google.com",
        "mx-inbound23-98.eu-central-1b.ess.aws.cudaops.com",
        "1e100.net",
        "notifications.google.com",
        "231.eu-central-1a.ess.aws.cudaops.com",
        "schema.org",
        "drive.google.com",
        "lh3.googleusercontent.com",
        "www.gstatic.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file2.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "mx.google.com",
        "example2.com",
        "egress-ip7a.ess.de.barracuda.com",
        "mail-ua1-f72.google.com",
        "mx-outbound13-247.eu-central-1a.ess.aws.cudaops.com",
        "1e100.net",
        "mail.example.com",
        "cloudscan13-198.eu-central-1a.ess.aws.cudaops.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file3.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "MWHPR12MB1837.namprd12.prod.outlook.com",
        "MN2PR12MB4062.namprd12.prod.outlook.com",
        "DM5PR12CA0055.namprd12.prod.outlook.com",
        "DM6NAM10FT006.eop-nam10.prod.protection.outlook.com",
        "DM5PR12CA0055.outlook.office365.com",
        "rapid7.com",
        "example.onmicrosoft.com",
        "protection.outlook.com",
        "mail-vs1-f41.google.com",
        "DM6NAM10FT006.mail.protection.outlook.com",
        "1e100.net",
        "mail.gmail.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file4.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "outlook.com",
        "office365.com",
        "microsoft.com",
        "onmicrosoft.com",
        "office.com",
        "xt.local",
        "eco.de"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file_without_subdomains.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "barracuda.com",
        "cudaops.com",
        "1e100.net",
        "schema.org",
        "googleusercontent.com",
        "gstatic.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file_without_subdomains2.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "example2.com",
        "barracuda.com",
        "cudaops.com",
        "1e100.net"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file_without_subdomains3.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "outlook.com",
        "office365.com",
        "rapid7.com",
        "onmicrosoft.com",
        "google.com",
        "1e100.net",
        "gmail.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_file_without_subdomains4.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "DM1PR09MB1111.namprd09.prod.outlook.com",
        "MN5PR09MB1111.namprd09.prod.outlook.com",
        "BN0PR06CA2222.namprd06.prod.outlook.com",
        "BN1NAM22FT333.eop-nam10.prod.protection.outlook.com",
        "BN0PR06CA2222.outlook.office365.com",
        "bounce.email2.microsoft.com",
        "example.onmicrosoft.com",
        "email2.microsoft.com",
        "protection.outlook.com",
        "test.email2.microsoft.com",
        "BN0PR06CA2222.mail.protection.outlook.com",
        "click.email2.office.com",
        "leave.email2.office.com",
        "7227040.xt.local",
        "eco.de",
        "atl1s07mta2695.xt.local",
        "BN0PR06CA2222.eop-nam06.prod.protection.outlook.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "mx.google.com",
        "drive-notifications.bounces.google.com",
        "egress-00000.ess.de.barracuda.com",
        "mail-aaa-bbbb.google.com",
        "mx-inbound23-98.eu-central-1b.ess.aws.cudaops.com",
        "1e100.net",
        "notifications.google.com",
        "231.eu-central-1a.ess.aws.cudaops.com",
        "schema.org",
        "drive.google.com",
        "lh3.googleusercontent.com",
        "www.gstatic.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string2.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "mx.google.com",
        "example2.com",
        "egress-ip7a.ess.de.barracuda.com",
        "mail-ua1-f72.google.com",
        "mx-outbound13-247.eu-central-1a.ess.aws.cudaops.com",
        "1e100.net",
        "mail.example.com",
        "cloudscan13-198.eu-central-1a.ess.aws.cudaops.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string3.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "MWHPR12MB1837.namprd12.prod.outlook.com",
        "MN2PR12MB4062.namprd12.prod.outlook.com",
        "DM5PR12CA0055.namprd12.prod.outlook.com",
        "DM6NAM10FT006.eop-nam10.prod.protection.outlook.com",
        "DM5PR12CA0055.outlook.office365.com",
        "rapid7.com",
        "example.onmicrosoft.com",
        "protection.outlook.com",
        "mail-vs1-f41.google.com",
        "DM6NAM10FT006.mail.protection.outlook.com",
        "1e100.net",
        "mail.gmail.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string4.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "outlook.com",
        "office365.com",
        "microsoft.com",
        "onmicrosoft.com",
        "office.com",
        "xt.local",
        "eco.de"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string_without_subdomains.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "barracuda.com",
        "cudaops.com",
        "1e100.net",
        "schema.org",
        "googleusercontent.com",
        "gstatic.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string_without_subdomains2.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "example.com",
        "google.com",
        "example2.com",
        "barracuda.com",
        "cudaops.com",
        "1e100.net"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string_without_subdomains3.json
</summary>
</details>

<details>

```
{
  "body": {
    "log": "rapid7/https://example.com Step name: domain_extractor\n",
    "meta": {},
    "output": {
      "domains": [
        "outlook.com",
        "office365.com",
        "rapid7.com",
        "onmicrosoft.com",
        "google.com",
        "1e100.net",
        "gmail.com"
      ]
    },
    "status": "ok"
  },
  "type": "action_event",
  "version": "v1"
}

```

<summary>
docker run --rm -i rapid7/extractit:2.2.1 --debug run < tests/domain_extractor_from_string_without_subdomains4.json
</summary>
</details>

<details>

```
[*] Validating plugin with all validators at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Executing validator ExceptionValidator
[*] Executing validator CredentialsValidator
[*] Executing validator PasswordValidator
[*] Executing validator PrintValidator
[*] Executing validator ConfidentialValidator
[*] Executing validator DockerValidator
[*] Executing validator URLValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "SupportedVersionValidator" failed! 
	Cause: supported_versions is missing from plugin.spec.yaml.


----
[*] Total time elapsed: 28649.067ms
```

<summary>
icon-validate --all .
</summary>
</details>

<details>

```
[*] Use ``make menu`` for available targets
[*] Including available Makefiles: ../../tools/Makefiles/Helpers.mk ../../tools/Makefiles/Colors.mk
--
[*] Running validators
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator CloudReadyConnectionCredentialTokenValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator EncodingValidator
[*] Executing validator ExampleInputValidator
[*] Executing validator CloudReadyValidator
[*] Executing validator SupportedVersionValidator
[*] Executing validator UnapprovedKeywordsValidator
[*] Executing validator HelpExampleValidator
[*] Plugin failed validation! The following validation errors occurred:

Validator "SupportedVersionValidator" failed! 
	Cause: supported_versions is missing from plugin.spec.yaml.


----
[*] Total time elapsed: 3251.426ms
```

<summary>
make validate
</summary>
</details>

### CODE REVIEW

File: plugins/extractit/icon_extractit/util/extractor.py [hunk 1]
```diff
@@ -32,16 +35,25 @@ def extract_filepath(provided_regex: str, provided_string: str, provided_file: s
...
         stripped_domain = tldextract.extract(match[1])
-        matches[match[0]] = ".".join(stripped_domain[1:3])
+        if not stripped_domain.suffix:
+            suffix = stripped_domain.domain
+            subdomain = stripped_domain.subdomain
+            if subdomain and suffix:
+                subdomains = subdomain.split(".")
+                matches[match[0]] = f"{subdomains[len(subdomains) - 1]}.{suffix}"
+        else:
+            matches[match[0]] = ".".join(stripped_domain[1:3])
```
Comment (mrinehart-r7) [inline plugins/extractit/icon_extractit/util/extractor.py:45]: Can you add some comments explaining the process here? It's a little difficult to follow this logic.

File: plugins/extractit/icon_extractit/util/util.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 class Regex:
-    Domain = r"(?:(?=[\p{L}\p{N}\p{Lo}\p{So}-]{1,63}\.)[\p{L}\p{N}\p{Lo}\p{So}]+(?:-[\p{L}\p{N}\p{Lo}\p{So}]+)*\.)+[\p{Ll}\p{Lo}]{2,63}(?:@|\/[\p{L}\p{N}@:%_\-+.~]*)?"
+    Domain = r"(?:(?=[\p{L}\p{N}\p{Lo}\p{So}-]{1,63}\.)[\p{L}\p{N}\p{Lo}\p{So}]+(?:-[\p{L}\p{N}\p{Lo}\p{So}]+)*\.)+[\p{Ll}\p{Lo}]{2,63}(?:@|=|\/[\p{L}\p{N}@:%_\/\-+?=#&.~]*)?"
```
Comment (mrinehart-r7) [inline plugins/extractit/icon_extractit/util/util.py:2]: Are you sure this regex works? I tried it in regex101 with Python regex and it was not picking up any domains.

Comment (dsliwinski-r7) [reply plugins/extractit/icon_extractit/util/util.py:2]: I ran a few more tests and the results are correct. Unit testing for the domain extractor is also passing, so surely this regex works. I also checked this in regex101 with the Python regex and it looks like it doesn't support using Unicode categories like \p{L} in the regex and therefore doesn't find any domains.

File: plugins/extractit/icon_extractit/util/extractor.py [hunk 3]
```diff
@@ -32,16 +35,28 @@ def extract_filepath(provided_regex: str, provided_string: str, provided_file: s
...
 def clear_domains(matches: list) -> list:
     new_matches = []
     for match in enumerate(matches):
         if not match[1].endswith("@"):
-            new_matches.append(match[1].split("/")[0])
-    return new_matches
+            split_match = match[1].split("/")[0]
+            if not split_match.endswith("="):
+                new_matches.append(split_match)
+    return list(dict.fromkeys(new_matches))
```
Comment (mberezin-r7) [inline plugins/extractit/icon_extractit/util/extractor.py:59]: If I understand this correctly, we are stripping away the parameters from the URL by looking for values with `=`. Could we not just strip away everything that comes after the `?` in the URL? We know that in a URL, parameters are anything that comes after the `?` symbol.
Do we want some of the domains in the parameters?
I think we could use some more comments & documentation explaining what this function does, thank you.

**Pull request merged**